Track precise size of FlatFileIndex, clean up tables

Swatinem opened the pull request:
This does a couple of things:
- It adds a new `size` field/column to precisely track the size of persisted FlatFileIndexes.
- Makes `ArtifactBundleIndex.release_name`, `ArtifactBundleIndex.dist_name` and `ArtifactBundleIndex.date_last_modified` nullable so we can remove them later on.

Swatinem commented on src/sentry/migrations/0545_flat_file_size_cleanup.py:
```diff
@@ -0,0 +1,97 @@
+                ),
+            ],
+        ),
+        # TODO? Remove the foreign key constraint from the legacy `flat_file_index_id` column
+        # migrations.AlterField(
+        #     model_name="artifactbundleflatfileindex",
+        #     name="flat_file_index",
+        #     field=sentry.db.models.fields.foreignkey.FlexibleForeignKey(
+        #         db_constraint=False, on_delete=models.deletion.CASCADE, to="sentry.File"
+        #     ),
+        # ),
```
Looks like I missed this step from the metadata migration in https://github.com/getsentry/sentry/pull/55764.
The docs only mention this step in the section about removing a whole table, so I’m unsure if it is strictly necessary, or if the docs need to be updated to highlight that it indeed is.

Swatinem replied:
Another confusing thing about this is that we still have a steady stream of queries for this, even though it has already been removed from the metadata a couple of days ago.

```
SELECT sentry_artifactbundleflatfileindex . id 
FROM sentry_artifactbundleflatfileindex 
WHERE sentry_artifactbundleflatfileindex . flat_file_index_id IN ( ? )
```

markstory replied:
That is really strange. Django shouldn't be querying fields that are not in the model. Could you be looking at older tracing data?

markstory commented on src/sentry/migrations/0549_flat_file_size_cleanup.py:
```diff
@@ -0,0 +1,74 @@
+            ],
+        ),
+        # Give fields a default so they can be removed in followup migrations
+        migrations.AlterField(
+            model_name="artifactbundleindex",
+            name="release_name",
+            field=models.CharField(default="", max_length=250),
+        ),
+        migrations.AlterField(
+            model_name="debugidartifactbundle",
+            name="source_file_type",
+            field=models.IntegerField(default=2),
+        ),
```
The SQL dump for this migration isn't showing any schema changes for these default values, as django likes to manage default values in the ORM. You would need to use `SeparateDatabaseAndState` to ensure DDL is applied. You can also make the fields nullable which will also generate DDL.

Swatinem replied:
jesus… ofc the existing `default`s of the columns are not in the DDL either but just in django land 🤦🏻 

So I changed this yet again to first set the columns as nullable, hopefully that will finally do the right thing.

Swatinem committed:
Track precise size of FlatFileIndex, clean up tables

This does a couple of things:
- It adds a new `size` field/column to precisely track the size of persisted FlatFileIndexes.
- Makes `ArtifactBundleIndex.release_name`, `ArtifactBundleIndex.dist_name` and `ArtifactBundleIndex.date_last_modified` nullable so we can remove them later on.

markstory commented on src/sentry/models/artifactbundle.py:
```diff
@@ -161,6 +161,7 @@ class ArtifactBundleFlatFileIndex(Model):
     project_id = BoundedBigIntegerField(db_index=True)
     release_name = models.CharField(max_length=250)
     dist_name = models.CharField(max_length=64, default=NULL_STRING)
+    size = BoundedPositiveIntegerField(default=0)
```
Should this be a BIGINT? Or are customer file sizes always going to fit in 32bits?

Swatinem replied:
Yes they would. But I’m not going to move forward with this PR as is anyway. Thanks for the feedback though.
